Introduce runtime config to run Mapbox animators on background thread

kiryldz opened the pull request:
<!--
Thanks for submitting a pull request!
Please fill out the sections below to complete your submission.
We appreciate your contributions!
-->

### Summary of changes

Introduce experimental runtime option to switch between main and background threads to run all Mapbox animators that include:
 - plugin-animation
 - plugin-locationcomponent
 - plugin-viewport

Switching to background thread is introduced as the workaround for Android Auto use-case on Xiaomi devices described in [this](https://github.com/mapbox/mapbox-maps-android/issues/1413) ticket. Mapbox `extension-androidauto` is also updated in this PR to switch to background thread __only__ when app goes to background on Xiaomi devices. All other devices are using main thread for animators as before.

Actual SDK behaviour is not affected by this change, main thread is still used to run all the animators.

No tests are written as part of this PR as map camera instrument tests behave slightly differently with background thread enabled in several tests and could be flaky; we do not wish to double instrument tests execution time as well. Perhaps instrument tests could be added as part of AA extension as tailwork. Main test in scope for this PR - make sure all existing tests are passing as before.

<!--
• If this is a new feature, include a short summary on how to use it.
• If this is a bug fix, explain how your contribution resolves the problem.
• Include a screenshot or gif if applicable
-->

### User impact (optional)

<!--
If this PR introduces user-facing changes, please note them here.
-->


## Pull request checklist:
 - [x] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
    <!--
        | Before | After |
        | ----- | ----- |
        | <img src="" width = 250/> | <img src="" width = 250/> |
        or
        | <video src="" width = 250/> | <video src="" width = 250/> |
    -->
 - [x] Write tests for all new functionality. If tests were not written, please explain why.
 - [ ] Optimize code for java consumption (`@JvmOverloads`, `@file:JvmName`, etc).
 - [ ] Add example if relevant.
 - [x] Document any changes to public APIs.
 - [x] Run `make update-api` to update generated api files, if there's public API changes, otherwise the `verify-api-*` CI steps might fail.
 - [x] Update [CHANGELOG.md](../CHANGELOG.md) or use the label 'skip changelog', otherwise `check changelog` CI step will fail.
 - [ ] If this PR is a `v10.[version]` release branch fix / enhancement, merge it to `main` firstly and then port to `v10.[version]` release branch.

Fixes: < Link to related issues that will be fixed by this pull request, if they exist >

PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-android/blob/main/CONTRIBUTING.md#contributor-license-agreement).

kmadsen commented on sdk-base/src/main/java/com/mapbox/maps/threading/AnimationThreadController.kt:
```diff
@@ -0,0 +1,112 @@
+   *
+   * @throws RuntimeException when called from non Android Main thread and [useBackgroundThread] was not called beforehand.
+   */
+  fun postOnAnimatorThread(function: () -> Unit) {
```
+1

kmadsen commented on sdk-base/src/main/java/com/mapbox/maps/threading/AnimationThreadController.kt:
```diff
@@ -0,0 +1,112 @@
+ *  only when it's needed.
+ */
+@MapboxExperimental
+object AnimationThreadController {
```
@kiryldz will you add tests for this?

i find objects easier to unit test if they have a internal class that is essentially identical

``` kotlin
object AnimationThreadController {
  private val delegate = AnimationThreadControllerDelegate()
  
  fun useBackgroundThread() = delegate.useBackgroundThread()

...
}
```

kiryldz replied:
first step will make sure that all existing tests are passing, tests for background will be added later separately. However thanks for the architectural tip, I will look into it

kmadsen commented on sdk-base/src/main/java/com/mapbox/maps/threading/AnimationThreadController.kt:
```diff
@@ -0,0 +1,112 @@
+ *
+ *  Switching between [useBackgroundThread] and [useMainThread] could be done at any point in runtime.
+ *
+ *  Using this class is extremely unsafe and error prone, it should be used in very specific use-cases!
+ *  Even when it's essential to use it (for example Android Auto use-case on specific devices when locking the device)
+ *  it would be safer to use Android Main thread as much as possible and switch to background thread
+ *  only when it's needed.
```
Agree here also the point is to make it safe and production ready. This will be the mechanism to ship production products to production xiaomi devices.

I think we should start a public issue which keeps track of known issues related to this class.

kiryldz replied:
let's do that when this will land - who knows, we could still hit some major blocker in this implementation

kmadsen commented:
Testing some and here's a crash

```
2022-09-01 11:30:35.604 14585-16472/com.mapbox.maps.testapp.auto E/AndroidRuntime: FATAL EXCEPTION: MapboxAnimThread
    Process: com.mapbox.maps.testapp.auto, PID: 14585
    java.lang.RuntimeException: Function 'postOnAnimatorThread' could not be used from non-main thread when 'useBackgroundThread()' was not called beforehand!
        at com.mapbox.maps.threading.AnimationThreadController.postOnAnimatorThread(AnimationThreadController.kt:89)
        at com.mapbox.maps.plugin.animation.animator.CameraAnimator.addListener(CameraAnimator.kt:102)
        at android.animation.AnimatorSet.addAnimationEndListener(AnimatorSet.java:1248)
        at android.animation.AnimatorSet.startAnimation(AnimatorSet.java:1189)
        at android.animation.AnimatorSet.start(AnimatorSet.java:729)
        at android.animation.AnimatorSet.start(AnimatorSet.java:684)
        at com.mapbox.maps.plugin.animation.CameraAnimationsPluginImpl$startHighLevelAnimation$2$1.invoke(CameraAnimationsPluginImpl.kt:903)
        at com.mapbox.maps.plugin.animation.CameraAnimationsPluginImpl$startHighLevelAnimation$2$1.invoke(CameraAnimationsPluginImpl.kt:903)
        at com.mapbox.maps.threading.AnimationThreadController.postOnAnimatorThread$lambda-2(AnimationThreadController.kt:85)
        at com.mapbox.maps.threading.AnimationThreadController.$r8$lambda$nE-XMEhVlqLA1g9jOyl1_QBh_Io(Unknown Source:0)
        at com.mapbox.maps.threading.AnimationThreadController$$ExternalSyntheticLambda0.run(Unknown Source:2)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loopOnce(Looper.java:210)
        at android.os.Looper.loop(Looper.java:299)
        at android.os.HandlerThread.run(HandlerThread.java:67)
```

kmadsen commented on android-auto-app/src/main/java/com/mapbox/maps/testapp/auto/car/MapSession.kt:
```diff
@@ -16,9 +16,10 @@ import com.mapbox.maps.extension.androidauto.mapboxMapInstaller
   private val carMapShowcase = CarMapShowcase()
   private val mapboxCarMap = mapboxMapInstaller()
-    .onCreated(CarMapWidgets(), carMapShowcase)
+    .onCreated(CarAnimationThreadController(), CarMapWidgets(), carMapShowcase)
```
Note that this is not part of the installer, because Android Automotive will not have this issue

kmadsen replied:
Although, maybe it should go inside anyways 🤔

kiryldz replied:
I can't advise anything here but in general better not use background thread unless there's absolute need for that imo

kmadsen replied:
think it's good like this for now. it's easy to add but not as easy to remove. resolving here

kiryldz commented:
> Testing some and here's a crash
> 
> ```
> 2022-09-01 11:30:35.604 14585-16472/com.mapbox.maps.testapp.auto E/AndroidRuntime: FATAL EXCEPTION: MapboxAnimThread
>     Process: com.mapbox.maps.testapp.auto, PID: 14585
>     java.lang.RuntimeException: Function 'postOnAnimatorThread' could not be used from non-main thread when 'useBackgroundThread()' was not called beforehand!
>         at com.mapbox.maps.threading.AnimationThreadController.postOnAnimatorThread(AnimationThreadController.kt:89)
>         at com.mapbox.maps.plugin.animation.animator.CameraAnimator.addListener(CameraAnimator.kt:102)
>         at android.animation.AnimatorSet.addAnimationEndListener(AnimatorSet.java:1248)
>         at android.animation.AnimatorSet.startAnimation(AnimatorSet.java:1189)
>         at android.animation.AnimatorSet.start(AnimatorSet.java:729)
>         at android.animation.AnimatorSet.start(AnimatorSet.java:684)
>         at com.mapbox.maps.plugin.animation.CameraAnimationsPluginImpl$startHighLevelAnimation$2$1.invoke(CameraAnimationsPluginImpl.kt:903)
>         at com.mapbox.maps.plugin.animation.CameraAnimationsPluginImpl$startHighLevelAnimation$2$1.invoke(CameraAnimationsPluginImpl.kt:903)
>         at com.mapbox.maps.threading.AnimationThreadController.postOnAnimatorThread$lambda-2(AnimationThreadController.kt:85)
>         at com.mapbox.maps.threading.AnimationThreadController.$r8$lambda$nE-XMEhVlqLA1g9jOyl1_QBh_Io(Unknown Source:0)
>         at com.mapbox.maps.threading.AnimationThreadController$$ExternalSyntheticLambda0.run(Unknown Source:2)
>         at android.os.Handler.handleCallback(Handler.java:938)
>         at android.os.Handler.dispatchMessage(Handler.java:99)
>         at android.os.Looper.loopOnce(Looper.java:210)
>         at android.os.Looper.loop(Looper.java:299)
>         at android.os.HandlerThread.run(HandlerThread.java:67)
> ```

Hard to say why it did happen, added several things like synchronized modifiers and cleaning handler's queue on exit - but I still could imagine this happening so transformed to an `logE` entry. Please keep an eye if that will pop up during testing and let me know!

kiryldz commented:
> > Testing some and here's a crash
> > ```
> > 2022-09-01 11:30:35.604 14585-16472/com.mapbox.maps.testapp.auto E/AndroidRuntime: FATAL EXCEPTION: MapboxAnimThread
> >     Process: com.mapbox.maps.testapp.auto, PID: 14585
> >     java.lang.RuntimeException: Function 'postOnAnimatorThread' could not be used from non-main thread when 'useBackgroundThread()' was not called beforehand!
> >         at com.mapbox.maps.threading.AnimationThreadController.postOnAnimatorThread(AnimationThreadController.kt:89)
> >         at com.mapbox.maps.plugin.animation.animator.CameraAnimator.addListener(CameraAnimator.kt:102)
> >         at android.animation.AnimatorSet.addAnimationEndListener(AnimatorSet.java:1248)
> >         at android.animation.AnimatorSet.startAnimation(AnimatorSet.java:1189)
> >         at android.animation.AnimatorSet.start(AnimatorSet.java:729)
> >         at android.animation.AnimatorSet.start(AnimatorSet.java:684)
> >         at com.mapbox.maps.plugin.animation.CameraAnimationsPluginImpl$startHighLevelAnimation$2$1.invoke(CameraAnimationsPluginImpl.kt:903)
> >         at com.mapbox.maps.plugin.animation.CameraAnimationsPluginImpl$startHighLevelAnimation$2$1.invoke(CameraAnimationsPluginImpl.kt:903)
> >         at com.mapbox.maps.threading.AnimationThreadController.postOnAnimatorThread$lambda-2(AnimationThreadController.kt:85)
> >         at com.mapbox.maps.threading.AnimationThreadController.$r8$lambda$nE-XMEhVlqLA1g9jOyl1_QBh_Io(Unknown Source:0)
> >         at com.mapbox.maps.threading.AnimationThreadController$$ExternalSyntheticLambda0.run(Unknown Source:2)
> >         at android.os.Handler.handleCallback(Handler.java:938)
> >         at android.os.Handler.dispatchMessage(Handler.java:99)
> >         at android.os.Looper.loopOnce(Looper.java:210)
> >         at android.os.Looper.loop(Looper.java:299)
> >         at android.os.HandlerThread.run(HandlerThread.java:67)
> > ```
> 
> Hard to say why it did happen, added several things like synchronized modifiers and cleaning handler's queue on exit - but I still could imagine this happening so transformed to an `logE` entry. Please keep an eye if that will pop up during testing and let me know!

Actually found an issue in the logic and removed log message as well, PTAL

SamuelBrucksch commented:
Hi @kiryldz 

what is the latest snaphot of this? I saw you added another change after the version I used for testing, can i get that as snapshot too?

kiryldz commented:
> Hi @kiryldz
> 
> what is the latest snaphot of this? I saw you added another change after the version I used for testing, can i get that as snapshot too?

It does not change any logic, just tests / changelogs / code refactor etc. No worries 😄 I'll tag you if I'll change something related to business logic.

kiryldz commented on sdk-base/src/main/java/com/mapbox/maps/threading/AnimationThreadController.kt:
```diff
@@ -0,0 +1,120 @@
+ *  only when it's needed.
+ */
+@MapboxExperimental
+object AnimationThreadController {
```
@kmadsen I agree and spent some time to prototype it however it complicates the approach dramatically. I don't think we need to do it now (we can add this later if needed) because:
1. `AnimationThreadController` is located in `sdk-base` module while `MapView`, `MapboxMap` etc are located in `sdk` module and `sdk-base` does not depend on `sdk`. We can't simply move `AnimationThreadController` to `sdk` module as our plugins depend on `sdk-base` and not `sdk`. Of course it still could be done but it's a bit of a burden.
2. Background thread should be used only for Android Auto + Xiaomi + main map gone to background use-case. For this case it does not actually matter that we change thread for all map instances as we do really care only about one. Using background thread in some other use-cases apart from this is high risk and does not provide any performance gain.

As said above - we could always revisit this later if this will be needed badly, all new APIs are experimental and we may break them if needed.

kmadsen replied:
> Background thread should be used only for Android Auto + Xiaomi + main map gone to background use-case.

This is a point that is a reason to bound the animation thread to android auto though. Right now as a singleton, the car will change both the car and phone (and watch one day?)

> AnimationThreadController is located in sdk-base module while MapView, MapboxMap etc are located in sdk module and sdk-base does not depend on sdk

Am unsure why this is a restriction. `MapView` and `MapboxMap` can see `AnimationThreadController` so they should be able to control the lifecycle 🤔 

> As said above - we could always revisit this later if this will be needed badly, all new APIs are experimental and we may break them if needed.

👍 I'm also onboard with moving forward with `AnimationThreadController` as a singleton for now.

The issue that we will face immediately though, is we need a stable api. For that we can move `CarAnimationThreadController` into the SDK and be able to declare it stable some time soon. It can surface the functionality and allow the `AnimationThreadController` change over time.

alexander-kulikovskii commented on plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt:
```diff
@@ -843,22 +880,27 @@ class CameraAnimationsPluginImpl : CameraAnimationsPlugin {
-            commitChanges()
+        // listeners in Android SDK use non thread safe lists
+        postOnAnimatorThread {
+          addListener(object : AnimatorListenerAdapter() {
```
maybe simplify because you've got only one function inside?

kiryldz replied:
don't get how to simplify it

alexander-kulikovskii commented on sdk-base/src/main/java/com/mapbox/maps/threading/AnimationThreadController.kt:
```diff
@@ -0,0 +1,120 @@
+  /**
+   * Indicates if background thread is used now to run Mapbox animators.
+   */
+  var usingBackgroundThread = false
```
private setter?

yunikkk commented on sdk-base/src/main/java/com/mapbox/maps/threading/AnimationThreadController.kt:
```diff
@@ -0,0 +1,120 @@
+        backgroundHandler.post { function.invoke() }
+      } ?: logW(TAG, "useBackgroundThread was called but handler is null, animator event is skipped!")
+    } else {
+      if (Looper.myLooper() == Looper.getMainLooper()) {
```
`postOnMainThread(function)` ?

yunikkk commented on plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/HighLevelAnimatorSet.kt:
```diff
@@ -10,6 +11,8 @@ internal data class HighLevelAnimatorSet(
   var started = false
```
Is this flag used anywhere? It seems not

kiryldz commented on sdk/src/main/java/com/mapbox/maps/MapboxMap.kt:
```diff
@@ -947,7 +948,9 @@ class MapboxMap :
    */
   override fun unproject(coordinate: MercatorCoordinate, zoomScale: Double): Point {
-    checkNativeMap("unproject")
+    if (!AnimationThreadController.usingBackgroundThread) {
```
removed thread checking for static core functions

yunikkk reviewed (approved):
LGTM

ank27 commented on sdk/src/main/java/com/mapbox/maps/MapboxMap.kt:
```diff
@@ -876,7 +876,6 @@ class MapboxMap :
    * @return Returns the distance measured in meters.
    */
   override fun getMetersPerPixelAtLatitude(latitude: Double, zoom: Double): Double {
-    checkNativeMap("getMetersPerPixelAtLatitude")
```
why these are removed ?

kiryldz replied:
Those are static gl-native methods used for calculation, no need to request them on main thread only. Additionally `unproject` method is used in custom evaluator `flyTo` implementation and will be always called on background thread (if animator started from background thread) without ability to re-post on main as it expects the result in the same callchain.

jush commented on sdk-base/src/main/java/com/mapbox/maps/threading/AnimationThreadController.kt:
```diff
@@ -0,0 +1,115 @@
+  @Synchronized
+  fun postOnMainThread(function: () -> Unit) {
+    if (Looper.myLooper() == Looper.getMainLooper()) {
+      function.invoke()
```
Since this method is marked as synchronized then it means that any other method calling `postOnMainThread` will block for as long as this `function.invoke` takes.

Is that ok? It sounds a bit risky to me to let external functions block this.

kiryldz replied:
thanks for the interesting offline discussion!
decided to remove synchronized and add volatile modifier for the bool flag

kmadsen commented on extension-androidauto/CHANGELOG.md:
```diff
@@ -9,6 +9,7 @@ Mapbox welcomes participation and contributions from everyone.
 * Add `Session.mapboxMapInstaller` and `Screen.mapboxMapInstaller` extension functions to create the installers. ([#1603](https://github.com/mapbox/mapbox-maps-android/pull/1603))
 
 ## Bug fixes 🐞
+* Fix map being stuck when app goes to background on Xiaomi devices. ([1638](https://github.com/mapbox/mapbox-maps-android/pull/1638))
```
This changelog is actually part of the main maps sdk right?

There is nothing to release for the `extension-androidauto`, but there is an experimental `AnimationThreadController` api that can be used as part of the next maps release

kiryldz replied:
got it, updated `README.md` file instead.

kmadsen replied:
Why wasn't it added to the changelog for maps? Experimental apis are not added to maps changelogs?

Thinking there may be multiple changes related to this, and it will be tracked by Maps sdk versions

kiryldz replied:
This experimental api is needed for one use-case only (AA) and it's not related to SDK code - thus no changelog entry was added. We don't want this feature be used in any SDK use-cases and that's why there's no changelog entry.

kmadsen commented on sdk-base/src/main/java/com/mapbox/maps/threading/AnimationThreadController.kt:
```diff
@@ -0,0 +1,112 @@
+import com.mapbox.maps.logW
+
+/**
+ * Experimental controller singleton allowing to use background thread for all Mapbox animations.
```
@kiryldz Will you add a row to the `README.md`

https://github.com/mapbox/mapbox-maps-android/tree/main/extension-androidauto#compatibility-with-maps-sdk-v10

| `AnimationThreadController` | :white_check_mark:  | v10.9+ |

or 

| Xiaomi animation support | :white_check_mark:  | v10.9+ |

kiryldz replied:
done

kmadsen replied:
Nit pick, but the ordering is not right.

The list of compatibility version are in descending order. This row was added to the top so the list is no longer sorted

kiryldz replied:
feel free to arrange this as needed in tailwork PRs
